HSBM.py

```diff
@@ -0,0 +1,442 @@
+        likelihood_ratio=None,
+    ):
+
+        _check_common_inputs(
+            min_components, max_components, cluster_kws, embed_kws=None
+        )
+
+        if cluster_method not in ["GMM", "KMeans", "Spherical-KMeans"]:
+            msg = "clustering method must be one of {GMM, Kmeans, Spherical-KMeans}"
+            raise ValueError(msg)
+
+        if delta_criter:
+            if delta_criter < 0:
+                raise ValueError("delta_criter must be positive")
+
+        if likelihood_ratio:
+            if likelihood_ratio <= 0 or likelihood_ratio >= 1:
+                raise ValueError("likelihood_ratio must be in (0,1)")
+
+        self.parent = parent
+        self.n_init = n_init
+        self.min_components = min_components
+        self.max_components = max_components
+        self.cluster_method = cluster_method
+        self.min_split = min_split
+        self.selection_criteria = selection_criteria
+        self.cluster_kws = cluster_kws
+        self.min_split = min_split
+        self.max_level = max_level
+        self.delta_criter = delta_criter
+        self.likelihood_ratio = likelihood_ratio
+        
+    def fit(self, X, adj=None):
+        self.fit_predict(X)
+        return self
+
+    def predict(self, X, adj=None):
+        check_is_fitted(self, ["model_"], all_or_any=all)
+        labels = self.fit_predict(X)
+        return labels
+    
+    def fit_predict(self, X, level=None, adj=None):
+        X = check_array(
+            X, dtype=[np.float64, np.float32],  ensure_min_samples=1
+        )
+        if level:
+            if not isinstance(level, int):
+                raise TypeError("level must be an int")
+
+        self.X = X
+        self.adj = adj
+
+        if self.max_components > X.shape[0]:
+            msg = "max_components must be >= n_samples, but max_components = "
+            msg += "{}, n_samples = {}".format(self.max_components, X.shape[0])
+            raise ValueError(msg)
+        
+        while True:
+            current_node = self._get_next_node()
+            if current_node:
+                current_node._cluster_and_decide()
+            else:
+                break
+
+        labels = self._to_labels()
+        if level:
+            labels = labels[:, level]
+
+        self.predictions = labels
+        return labels
+
+    def _get_next_node(self):
+        leaves = self.root.leaves
+        current_node = []
+
+        for leaf in leaves:
+            if (len(leaf.X) >= self.max_components and
+                    len(leaf.X) >= self.min_split and
+                    leaf.depth < self.max_level):
+                if not hasattr(leaf, "k_"):
+                    current_node = leaf
+                    break
+        return current_node
+
+    def _cluster_and_decide(self):
+        X = self.X
+        if self.cluster_method == "GMM":
+            cluster = AutoGMMCluster(
+                min_components=1, max_components=self.max_components, **self.cluster_kws
```

Comment:
should this be `min_components`?

Reply:
Sorry I don't follow. Which line is it?

Reply:
line 147 you always set min_components to 1, is that desired?

Reply:
I'm setting it as 1 because I don't really see why someone would want a higher threshold for it: 
i) if that's not together w/ settings on max_level or min_split, the algorithm will keep splitting (until too few samples left); 
ii) even if it is indeed, I think the person would have a very clear vision for the data and really want the original data (& all following sub-clusters) to be recursively clustered until a specific level or size.

But I'm kind of hesitant about doing this for level 0 as well. Maybe I should change it so that it allows customized min_comp at level 0. 

And maybe there are other cases where someone does have the need to customize I didn't think of. What do you imagine?